incubator/hnc/internal/validators/namespace.go

```diff
@@ -64,48 +64,73 @@ func (v *Namespace) Handle(ctx context.Context, req admission.Request) admission
+		// allowing setting a parent when validating the HierarchyConfiguration.
+		if rsp := v.conflictBetweenParentAndExternalManager(req, ns); !rsp.Allowed {
+			return rsp
+		}
+	case v1beta1.Delete:
+		if rsp := v.cannotDeleteSubnamespace(req); !rsp.Allowed {
+			return rsp
+		}
+		if rsp := v.illegalCascadingDeletion(ns); !rsp.Allowed {
+			return rsp
 		}
 	}
 
-	// Allow all update namespace requests except if the namespace has a parent and wants
-	// to add "hnc.x-k8s.io/managedBy" annotation other than "hnc.x-k8s.io".
-	if req.op == v1beta1.Update && req.ns.Annotations[api.AnnotationManagedBy] != "" {
-		if req.ns.Annotations[api.AnnotationManagedBy] != api.MetaGroup && ns.Parent() != nil {
-			msg := fmt.Sprintf("Namespace %q is a child of %q. Namespaces with parents defined by HNC cannot also be managed externally. "+
-				"To manage this namespace with %q, first make it a root in HNC.", ns.Name(), ns.Parent().Name(), req.ns.Annotations[api.AnnotationManagedBy])
-			return deny(metav1.StatusReasonForbidden, msg)
+	return allow("")
+}
+
+func (v *Namespace) nameExistsInExternalHierarchy(req *nsRequest) admission.Response {
+	for _, nm := range v.Forest.GetNamespaceNames() {
+		if _, ok := v.Forest.Get(nm).ExternalTreeLabels[req.ns.Name]; ok {
+			msg := fmt.Sprintf("The namespace name %q is reserved by the external hierarchy manager %q.", req.ns.Name, v.Forest.Get(nm).Manager)
+			return deny(metav1.StatusReasonAlreadyExists, msg)
 		}
 	}
+	return allow("")
+}
+
+func (v *Namespace) conflictBetweenParentAndExternalManager(req *nsRequest, ns *forest.Namespace) admission.Response {
+	mgr := req.ns.Annotations[api.AnnotationManagedBy]
+	if mgr != "" && mgr != api.MetaGroup && ns.Parent() != nil {
+		msg := fmt.Sprintf("Namespace %q is a child of %q. Namespaces with parents defined by HNC cannot also be managed externally. "+
+			"To manage this namespace with %q, first make it a root in HNC.", ns.Name(), ns.Parent().Name(), mgr)
+		return deny(metav1.StatusReasonForbidden, msg)
+	}
+	return allow("")
+}
+
+func (v *Namespace) cannotDeleteSubnamespace(req *nsRequest) admission.Response {
+	parent := req.ns.Annotations[api.SubnamespaceOf]
+	if parent != "" {
+		msg := fmt.Sprintf("The namespace %s is a subnamespace. Please delete the anchor from the parent namespace %s to delete the subnamespace.", req.ns.Name, parent)
+		return deny(metav1.StatusReasonForbidden, msg)
+	}
+	return allow("")
+}
 
-	// Early exit to allow other non-delete requests or if the namespace allows cascading deletion.
-	if req.op != v1beta1.Delete || ns.AllowsCascadingDelete() {
+func (v *Namespace) illegalCascadingDeletion(ns *forest.Namespace) admission.Response {
+	// Early exit if the namespace allows cascading deletion.
+	if ns.AllowsCascadingDelete() {
 		return allow("")
 	}
 
-	// Check if the deleting namespace has subnamespaces that can't be deleted.
+	// Check if all children allow cascading deletion.
```

Comment:
Maybe file an issue to consider removing this exception? It made sense back when we required allowCascadingDelete to delete a leaf namespace as well, but the new interpretation of aCD now seems to be "will this operation delete _any_ namespace other than the one explicitly named?"

Reply:
sg will test first and then file an issue.

Reply:
So i think we still need to keep it and it's a right behavior. Here's the test I did.
```
k create ns a
k hns create b -n a
k delete ns a
```
It shouldn't allow it because the user may not aware it would also delete ns b. Currently, with this rule, the webhook will deny the deletion. WDYT?

Reply:
Basically, we should allow `k delete subns b -n a` or ( `k hns set a -a` + `k delete ns a` ), but not `k delete ns a` directly. WDYT?

Reply:
Yeah I was talking about making the webhook _more_ strict, not _less_. We can discuss it in another issue, but for now the behaviour of the webhook is correct.

Reply:
Thanks for the clarification, filed #876 for this.